File: bench/benches/sort.rb
require_relative "../lib/bench"

# sort N pseudo-random ints (MINSTD LCG), order-dependent rolling-hash checksum.
N = 5000

def work
  x = 1
  data = []
  N.times { x = (16807 * x) % 2147483647; data << x }
  data.sort!
  h = 0
  data.each { |v| h = (h * 31 + v) % 1000000007 }
  h
end

bench("sort") { work }

File: bench/benches/sort.clj
;; sort N pseudo-random ints (MINSTD LCG), order-dependent rolling-hash checksum.
(load-file "lib/bench.clj")
(def N 5000)
(defn gen [n]
  (loop [i 0 x 1 acc []]
    (if (< i n)
      (let [nx (mod (* 16807 x) 2147483647)] (recur (inc i) nx (conj acc nx)))
      acc)))
(defn hsh [coll]
  (reduce (fn [h v] (mod (+ (* h 31) v) 1000000007)) 0 coll))
(bench "sort" (fn [] (hsh (sort (gen N)))))
